pkg/route/secret/manager.go

```diff
@@ -0,0 +1,146 @@
+package secret
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	routev1 "github.com/openshift/api/route/v1"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+type Manager struct {
+	monitor SecretMonitor
+	// map of handlerRegistrations
+	registeredHandlers map[ObjectKey]SecretEventHandlerRegistration
+
+	lock sync.RWMutex
+
+	// monitors are the producer of the resourceChanges queue
+	resourceChanges workqueue.RateLimitingInterface
+
+	secretHandler cache.ResourceEventHandlerFuncs
+}
+
+func NewManager(kubeClient *kubernetes.Clientset, queue workqueue.RateLimitingInterface) *Manager {
+	return &Manager{
+		monitor:            NewSecretMonitor(kubeClient),
+		lock:               sync.RWMutex{},
+		resourceChanges:    queue,
+		registeredHandlers: make(map[ObjectKey]SecretEventHandlerRegistration),
+
+		// default secret handler
+		secretHandler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) {},
+			UpdateFunc: func(oldObj, newObj interface{}) {},
+			DeleteFunc: func(obj interface{}) {},
+		},
+	}
+}
+
+func (m *Manager) WithSecretHandler(handler cache.ResourceEventHandlerFuncs) *Manager {
+	m.secretHandler = handler
+	return m
+}
+
+func (m *Manager) RegisterRoute(parent *routev1.Route) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	var secretName string
+
+	if parent.Spec.TLS.ExternalCertificate != nil && len(parent.Spec.TLS.ExternalCertificate.Name) > 0 {
+		secretName = parent.Spec.TLS.ExternalCertificate.Name
+	} else {
+		return apierrors.NewInternalError(fmt.Errorf("unable to get secret name from route %v", parent))
+	}
+
+	key := generateKey(parent.Namespace, parent.Name, secretName)
+
+	handlerRegistration, err := m.monitor.AddSecretEventHandler(key.Namespace, key.Name, m.secretHandler)
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+	m.registeredHandlers[key] = handlerRegistration
+	klog.Info("secret manager registered route ", key)
+
+	return nil
+}
+
+func (m *Manager) UnregisterRoute(parent *routev1.Route) error {
```

Comment:
deletes may only have namespace/name if the final state is unknown.  You must be able to remove with nothing but a route namespace and name.  This can definitely be done, but may require another instance variable.

Reply:
Thanks for the suggestion, I've done some refinement and now it only needs namespace and routeName.

```go
func (m *Manager) UnregisterRoute(namespace, routeName string) error
```